x-pack/auditbeat/module/system/packages/rpm_linux.go

```diff
@@ -0,0 +1,303 @@
+//  rpmdbFreeIterator = (rpmdbMatchIterator (*)(rpmdbMatchIterator))f;
+//
+//  rpmdbFreeIterator(mi);
+//}
+//
+//void
+//my_rpmtsFree(void *f, rpmts ts) {
+//  rpmts (*rpmtsFree)(rpmts);
+//  rpmtsFree = (rpmts (*)(rpmts))f;
+//
+//  rpmtsFree(ts);
+//}
+import "C"
+
+// Constants in sync with /usr/include/rpm/rpmtag.h
+const (
+	RPMTAG_NAME        = 1000
+	RPMTAG_VERSION     = 1001
+	RPMTAG_RELEASE     = 1002
+	RPMTAG_SUMMARY     = 1004
+	RPMTAG_LICENSE     = 1014
+	RPMTAG_URL         = 1020
+	RPMTAG_ARCH        = 1022
+	RPMTAG_SIZE        = 1009
+	RPMTAG_INSTALLTIME = 1008
+)
+
+type cFunctions struct {
+	rpmtsCreate        unsafe.Pointer
+	rpmReadConfigFiles unsafe.Pointer
+	rpmtsInitIterator  unsafe.Pointer
+	rpmdbNextIterator  unsafe.Pointer
+	headerLink         unsafe.Pointer
+	headerGetEntry     unsafe.Pointer
+	headerFree         unsafe.Pointer
+	rpmdbFreeIterator  unsafe.Pointer
+	rpmtsFree          unsafe.Pointer
+}
+
+var cFun *cFunctions
+
+func dlopenCFunctions() (*cFunctions, error) {
+	var librpmNames = []string{
+		"/usr/lib64/librpm.so",
+	}
+	var cFun cFunctions
+
+	librpm, err := dlopen.GetHandle(librpmNames)
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.rpmtsCreate, err = librpm.GetSymbolPointer("rpmtsCreate")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.rpmReadConfigFiles, err = librpm.GetSymbolPointer("rpmReadConfigFiles")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.rpmtsInitIterator, err = librpm.GetSymbolPointer("rpmtsInitIterator")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.rpmdbNextIterator, err = librpm.GetSymbolPointer("rpmdbNextIterator")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.headerLink, err = librpm.GetSymbolPointer("headerLink")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.headerGetEntry, err = librpm.GetSymbolPointer("headerGetEntry")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.headerFree, err = librpm.GetSymbolPointer("headerFree")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.rpmdbFreeIterator, err = librpm.GetSymbolPointer("rpmdbFreeIterator")
+	if err != nil {
+		return nil, err
+	}
+
+	cFun.rpmtsFree, err = librpm.GetSymbolPointer("rpmtsFree")
+	if err != nil {
+		return nil, err
+	}
+
+	return &cFun, nil
+}
+
+func listRPMPackages() ([]*Package, error) {
+	// TODO; cache these handles
+
+	if cFun == nil {
+		var err error
+		cFun, err = dlopenCFunctions()
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	rpmts := C.my_rpmtsCreate(cFun.rpmtsCreate)
+	if rpmts == nil {
+		return nil, fmt.Errorf("Failed to get rpmts")
+	}
+	defer C.my_rpmtsFree(cFun.rpmtsFree, rpmts)
+	res := C.my_rpmReadConfigFiles(cFun.rpmReadConfigFiles)
+	if int(res) != 0 {
+		return nil, fmt.Errorf("Error: %d", int(res))
+	}
+
+	mi := C.my_rpmtsInitIterator(cFun.rpmtsInitIterator, rpmts)
+	if mi == nil {
+		return nil, fmt.Errorf("Failed to get match iterator")
+	}
+	defer C.my_rpmdbFreeIterator(cFun.rpmdbFreeIterator, mi)
+
+	packages := []*Package{}
```

Comment:
Use `var packages []*Package` to avoid allocating the zero length slice. https://github.com/golang/go/wiki/CodeReviewComments#declaring-empty-slices